### ISSUE TITLE
manage external link type admin page  E597

### ISSUE THREAD

**Issue by only1chunts** [gigascience/gigadb-website#299]
## User Story

>As a curator
>I want an admin page that will allow me to create new tabs for external links
>So that external links can be diplayed in the tabs of my choosing on a dataset page

## Acceptance criteria

>Given I have a new type of link i want to add to GigaDB
>When I go the admin page for external links
>Then I can add the new external_link name
> And I can add a description of that external link-type
>And I can tick a box to determine if there can be multiple instances of that external_link per dataset (default is yes multiple is allowed)
> And I can add a prefix of the URL if desired (e.g. https://doi.org/)
> And I can select whether that external_link type is displayed as a link or as a tab

>Given I want to update the name of a type of external link
>When I go the admin page for external links
>And select the type I want to change
>And update the name
>Then the name of the type is udpated in the external_link_type table

>Given I want to update the description of a type of external link
>When I go the admin page for external links
>And select the type I want to change
>And update the description
>Then the description of the type is udpated in the external_link_type table

repeat the above updates block for each of the things that can be updated (name, description, multiple, prefix, tab vs link)

>Given a new external link type is created as tab
>When I use that type in a dataset
>And I navigate to the dataset page
>then I see the tab after all the other external link types tab but before the history tab

## Additional infos

we would like to be able to add multiple new tabs to the dataset view using iframes. i.e. behave in the same way the jbrowse iframe works, just for other websites/links.
To do that we need an admin page to be able to define the tab type (external link type) which gives us the ability to name the tab, and provide an (optional) url prefix. e.g. for a GigaScience manuscript we could define the Tab name as "Linked Manuscript" the prefix could be "https://doi.org/".
It would also be useful to have an option here to allow multiple links in a single window as we do for protocols.io. I.e. another tickbox in the manage external link type admin page to assign occurrences 0=allow multiple (like protocols.io), 1=single entity.
Each tab will eventually have an "i" information icon on it to tell users what the tab is showing, so each external link type should also have a definition that can be shown by that "i". These definitions will also need to be able to be added/edited in the manage external link type admin page 

Currently the new Tabs are required for;
Workflows  (CWL viewer or WorkFlowHub views)
DOME registry link
Resource PIDs (BioTools or RRIDs)

This Story is part of Epic #597

**Comment by alli83** [comment]
@only1chunts if I understand correctly, we need to add more columns to the table for external links type and be able to update the name right?

**Comment by only1chunts** [comment]
There is currently a database table called `external_link_type` which has the columns `id` and `name`, (this is straying into the realms of implementation so I may be wrong...) I think we need 3 additional columns there to cover the criteria in;
 
> Given I have a new type of link i want to add to GigaDB
> When I go the admin page for external links
> Then I can add the new external_link name
> And I can add a description of that external link-type
> And I can tick a box to determine if there can be multiple instances of that external_link per dataset (default is yes multiple is allowed)
> And I can add a prefix of the URL if desired (e.g. https://doi.org/)
> And I can select whether that external_link type is displayed as a link or as a tab

those columns being:
 `link_type_description` - to hold the description of each external-link-type
`multiple-values-allowed` - to hold the yes/no option of whether multiple values can be included per dataset
`URL-prefix` - to hold the optional prefix value ,eg. https://doi.org/
`display-as-tab` - to tell the user interface whether to display these external-link-types as a tab or as URL links

Its possible the implementation will need to be different, but you will need to look at the behaviour each of the current external-link-types to see if those could be accommodated by the above? i.e. currently everytime we want to add a new external-link-type the tech team have to do it bespoke, we want to remove that headache so that curators can add them in using a data-admin page for the external_link_type table (there is no admin page for that yet, hopefully its covered under a different ticket in E597?)

It maybe that there is a need for another column to enable us to say under which relationship those links should be added to the DataCite metadata?

**Comment by alli83** [comment]
@only1chunts thanks, I was thinking the same. Appreciate the confirmation!

**Comment by alli83** [comment]
@only1chunts if you need "It maybe that there is a need for another column to enable us to say under which relationship those links should be added to the DataCite metadata?", I can do it

**Comment by alli83** [comment]
@only1chunts would it be possible to get a mockup to better understand what you mean by 'displayed as tab or link'?

**Comment by only1chunts** [comment]
@alli83 if you think it will be helpful to have that information to use in the datacite metadata generation (next time we do any updates to that) then yes please add the field now

The values for the field should be constrained by the datacite vocabulary for relationshipTypes (I think you already have those in the code base here data/production_like/relationship.csv ). 

The field should include a description/help button to aid curators in knowing what to put in it, e.g. "Specify the relationship the values entered in this field will have with the dataset(s) they are being linked to, e.g.  a protocol link would be related to the dataset with the relationshipType `references`"

**Comment by only1chunts** [comment]
![Image](https://github.com/user-attachments/assets/b889560d-c9a9-4483-8b31-af06d20badb0)
Does that help?

**Comment by alli83** [comment]
@only1chunts yes, thanks! so if displayed as tab I will need to create a tab 'links'?

**Comment by only1chunts** [comment]
no the tab will be called whatever the new link is named by the person adding it, e.g. If I wanted to add a new external_link_type for "Workflows" then I would complete the fields on the admin form as:
(ID= auto assigned)
Name = Workflow
link_type_description = The workflow(s) associated with the dataset.
multiple-values-allowed = yes
URL-prefix = na
display-as-tab = yes
relationshipType = references

The expectation would then be that a new tab be created in any dataset using this `external_link_type` with the heading Workflow, and it would display the content of the URL given as the value in the for that dataset in the `external_link` table

**Comment by alli83** [comment]
@only1chunts ok thanks!

**Comment by alli83** [comment]
@only1chunts do you only need "Additional information", "Genome browser", "Github links" external link type or all of them? It seems the there is currently a limit on the displayed types

**Comment by alli83** [comment]
@only1chunts "Protocols.io", "JBrowse", "3D Models", "Code Ocean" are displayed as tab (before my PR for this issue). Should we keep it or now it will only depend on the new property "as tab or link"?

**Comment by only1chunts** [comment]
I'm guessing we will need to retrofit all existing external_link_types to the new table layout, so I'll try to specify all the values in a table and add it here shortly.

**Comment by alli83** [comment]
@only1chunts ok!

**Comment by alli83** [comment]
@only1chunts it will be possible to update them via one of the form I just created for this PR.

**Comment by only1chunts** [comment]
| Name | link_type_description | multiple-values-allowed | URL-prefix | display-as-tab | relationshipType
-- | -- | -- | -- | -- | -- | --
existing | Additional information | URL links to data resources that are to be considered part of the dataset | yes | na | Link | references
existing | Genome Browser | A link to an externally hosted genome browser displaying the data   included in this dataset | no | na | Link | references
existing | Protocols.io | the DOI of the individual protocols or the protocol collection showing   the methods used in this dataset | yes | na | Link | references
existing | Jbrowse | the GigaDB hosted Jbrowse link displaying the genome of the dataset | no | na | Tab | hasPart
existing | 3D models | The link to the 3D model(s) hosted in this dataset | yes | na | Tab | hasPart
existing | Code Ocean | The Code Ocean DOI of the code utilised in this dataset | yes | na | Tab | references
existing | GitHub Links | The GitHub repository URL of any software/code hosted in GitHub that is   to be included as part of this dataset | yes | na | Link | references
existing | UCSC Tumour Map Viewer | A link to the UCSC tumour map viewer | no | na | Link | references
new | 3D   SketchFab | URL of the 3D model(s) from this dataset, hosted in SketchFab | yes | na | Tab | references
new | Software archive | The Software heritage archive (or similar) PID for the archival version   of record of the code included in this dataset. | yes | na | link | references
new | Workflow | The workflow(s) associated with the dataset | yes | na | tab | references
new | Resource ID | The Research Resource Identifiers (RRID) of the resource(s) that are to be considered part of the dataset | yes | https://scicrunch.org/resolver/ | Link | References 
new | Referenced GitHub Links| The GitHub repository URL of any software/code reused by dataset (i.e. not created for this dataset) | yes | na | Link | references

**Comment by alli83** [comment]
@only1chunts thanks! I will create a migration file and update them

**Comment by alli83** [comment]
@only1chunts just a filter on the admin page for name  and description, is that enough for you?

**Comment by only1chunts** [comment]
yes that should suffice, thanks. Hopefully we'll never have soo many different external_link_types that have much need for a filter, but as you're offering its a nice bonus 👍

**Comment by rija** [comment]
Hi @only1chunts, @alli83 
minor correction: migration for 3D Sketchfab is already existing in the main branch as it is in release 4.4.7 currently on staging.
@alli83, if you rebase your PR onto the head of develop, you will get it and can build one top of it

**Comment by alli83** [comment]
@only1chunts could you specify for 
'Pre-Print', 'Referenced GitHub repo' and  'RRID' please?

**Comment by only1chunts** [comment]
The pre-print we agreed would be dealt with in the manuscript table not the external link table (see ticket #2236)
I have added the "Referenced GitHub Repo" and "RRID" details in the table above.

**Comment by alli83** [comment]
> The pre-print we agreed would be dealt with in the manuscript table not the external link table (see ticket [#2236](https://github.com/gigascience/gigadb-website/issues/2236)) I have added the "Referenced GitHub Repo" and "RRID" details in the table above.

@only1chunts yes indeed! I already worked on the manuscript. I thought there was also a pre print external link type but it's was just a misunderstanding on my part :)

**Comment by alli83** [comment]
@only1chunts are 'Software archive' and 'Software Heritage Archive (SWHA) the same?

**Comment by alli83** [comment]
@only1chunts just to be sure, because we have different information on each ticket: Apparently it's not like in the above table "github links" and "Referenced GitHub Links" but "Authors code repositories" and "Cited code repository" (from https://github.com/gigascience/gigadb-website/issues/1137). is it correct?

**Comment by only1chunts** [comment]
Correct, software heritage library provide software archiving service, but
there are others that offer the service so we should keep it generic by
name, i.e. call the link software archive.

**Comment by only1chunts** [comment]
I'd forgotten about that ticket, but yes I prefer the more generic names of
authors code repo and cited code repo. That doesn't restrict us to just
GitHub then 👍

**Comment by pli888** [comment]
Closing this ticket because there is PR #2253 for this work.

**Comment by alli83** [comment]
the PR is still opened https://github.com/gigascience/gigadb-website/pull/2253
